emacs/.doom.d/config.org
# -*- lexical-binding: t; -*-
#+OPTIONS: num:nil
#+SETUPFILE: ~/devel/ext/org-html-themes/org/theme-readtheorg-local.setup
#+PROPERTY: header-args :noweb-prefix no :noweb no-export :eval no
#+PROPERTY: header-args:elisp :tangle config.el
#+PROPERTY: header-args:scheme :tangle no :exports code
#+PROPERTY: header-args:cfg :tangle no :eval yes :cache yes :exports code :wrap "src text :noweb-ref guix-file"
#+begin_src elisp :exports none
;; -*- lexical-binding: t; -*-
;; generated from config.org
(require 's)
(require 'f)
#+end_src
#+begin_src elisp :tangle packages.el :exports none
;; generated from config.org
#+end_src
* channels
** guix
#+begin_src scheme :noweb-ref guix-channel
(channel
 (name 'guix)
 (url "https://git.guix.gnu.org/guix.git")
 (branch "master")
 (introduction
  (make-channel-introduction
   "9edb3f66fd807b096b48283debdcddccfea34bad"
   (openpgp-fingerprint
    "BBB0 2DDF 2CEA F6A8 0D1D  E643 A2A0 6DF2 A33A 54FA"))))
#+end_src
** nonguix
#+begin_src scheme :noweb-ref guix-channel
(channel
 (name 'nonguix)
 (url "https://gitlab.com/nonguix/nonguix")
 (branch "master")
 (introduction
  (make-channel-introduction
   "897c1a470da759236cc11798f4e0a5f7d4d59fbc"
   (openpgp-fingerprint
    "2A39 3FFF 68F4 EF7A 3D29  12AF 6F51 20A0 22FB B2D5"))))
#+end_src
** COMMENT sops
I use only the sops package which is in main guix channel now.
#+begin_src scheme :noweb-ref guix-channel
(channel
 (name 'sops-guix)
 (url "https://github.com/fishinthecalculator/sops-guix.git")
 (branch "main")
 (introduction
  (make-channel-introduction
   "0bbaf1fdd25266c7df790f65640aaa01e6d2dbc9"
   (openpgp-fingerprint
    "8D10 60B9 6BB8 292E 829B  7249 AED4 1CC1 93B7 01E2"))))
#+end_src
** personal
#+begin_src scheme :noweb-ref guix-channel
(channel
 (name 'personal)
 (url "https://github.com/sarg/dotfiles")
 (branch "master"))
#+end_src
* globals
#+begin_src scheme :noweb-ref globals
'((name . "Sergey Trofimov")
  (email . "sarg@sarg.org.ru")
  (gpgKey . "3ADB423B40A20785")
  (unlock-pam-service . "fingerprint"))
#+end_src
#+begin_src elisp
(defvar guix/globals <<globals>>)
(defvar sarg/media-dir "/storage/media")
#+end_src
* system.scm
** template
#+begin_src scheme :tangle ../../guix/system.scm
(use-modules (gnu)
             (gnu artwork)
             (gnu services)
             (gnu bootloader grub)
             (gnu system privilege)
             (gnu system pam)
             (gnu packages)
             (guix packages)
             (guix channels)
             (guix utils)
             (srfi srfi-1)
             (personal services utils)
             (nongnu system linux-initrd)
             (nongnu packages linux)
             (ice-9 receive)
             <<system-use>>)

(define (pkg x)
  (if (string? x)
      (receive (package out)
          (specification->package+output x)
        (if (string= "out" out) package
            (list package out)))
      x))

(define (auto-login-to-tty config tty user)
  (if (string=? tty (mingetty-configuration-tty config))
      (mingetty-configuration
       (inherit config)
       (auto-login user)
       (shepherd-requirement
        (cons 'dbus-system
              (mingetty-configuration-shepherd-requirement
               config))))
      config))

(operating-system
  (kernel linux)
  ; https://bbs.archlinux.org/viewtopic.php?id=314578
  (kernel-arguments '("quiet" "loglevel=1" "amdgpu.dcdebugmask=0x40000"))
  (initrd microcode-initrd)
  (firmware (cons* amdgpu-firmware realtek-firmware %base-firmware))
  (locale "C.UTF-8")
  (timezone "Europe/Berlin")
  (host-name "elitebook")
  (keyboard-layout (keyboard-layout "us"))
  (name-service-switch
    (mdns-host-lookup-nss #:ipv6? #f))
  (bootloader
   (bootloader-configuration
    (theme (grub-theme
            (image (file-append %artwork-repository "/grub/guix-16-9.svg"))
            (resolution '(1920 . 1080))
            (gfxmode '("1920x1080"))))
    (bootloader grub-efi-bootloader)
    (targets '("/boot"))))

  (file-systems
   (cons* (file-system
            (mount-point "/")
            (device (file-system-label "Guix_image"))
            (type "ext4"))
          (file-system
            (mount-point "/boot")
            (device (file-system-label "GNU-ESP"))
            (type "vfat"))
          %base-file-systems))

  (users (cons* (user-account
                 (name "sarg")
                 (comment "Sergey Trofimov")
                 (group "users")
                 (home-directory "/home/sarg")
                 (supplementary-groups
                  '("wheel" "netdev" "audio" "video" "tty" "input" "cdrom" "dialout"
                    <<user-group>>)))
                %base-user-accounts))

  (packages
   (append
    (map pkg (list <<system-pkg>>))
    (filter (lambda (p)
              (not (member (package-name p)
                           '("wireless-tools" "info-reader" "nano" "mg" "isc-dhcp"))))
            %base-packages)))

  (pam-services (cons* <<pam-service>> (base-pam-services)))
  (privileged-programs
   (cons* <<privileged-program>> %default-privileged-programs))

  (services
   (cons*
    <<system-service>>
    (modify-services %base-services
      <<modify-system-service>>))))
#+end_src
** COMMENT greetd
#+begin_src scheme :noweb-ref system-use
(guix build-system trivial)
#+end_src
#+begin_src scheme :noweb-ref system-pkg
(package
  (name "emacs-ewm-desktop")
  (version "0")
  (source #f)
  (build-system trivial-build-system)
  (arguments
   (list
    #:builder
    (with-imported-modules '((guix build utils))
      #~(begin
          (use-modules (guix build utils))
          (make-desktop-entry-file
           (string-append #$output "/share/wayland-sessions/ewm.desktop")
           #:name "EWM"
           #:type "Application"
           #:exec "emacs --fg-daemon -f ewm-start-module"
           #:comment
           '((#f "Emacs Wayland Manager")))))))
  (license #f) (home-page #f) (synopsis #f) (description #f))
#+end_src
#+begin_src scheme :noweb-ref modify-system-service
(delete login-service-type)
(delete mingetty-service-type)
(delete console-font-service-type)
#+end_src
#+begin_src scheme :noweb-ref system-service
(service greetd-service-type
        (greetd-configuration
         (greeter-supplementary-groups '("input"))
         (terminals
          (list (greetd-terminal-configuration
                 (terminal-vt "1"))
                (greetd-terminal-configuration
                 (terminal-vt "2")
                 (terminal-switch #t)
                 (default-session-command
                   (file-append (pkg "tuigreet") "/bin/tuigreet")))
                (greetd-terminal-configuration
                 (terminal-vt "3"))))))
#+end_src
** base
#+begin_src scheme :noweb-ref modify-system-service
(mingetty-service-type
 config =>
 (auto-login-to-tty config "tty1" "sarg"))

(guix-service-type
 config =>
 (guix-configuration
  (inherit config)
  ;; prevent guix gc deleting build inputs
  (privileged? #f)
  (extra-options '("--gc-keep-derivations"
                   "--gc-keep-outputs"))))
#+end_src
** nonguix
#+name: cfg-nonguix-pub
#+begin_src cfg :wrap src text
(public-key (ecc (curve Ed25519) (q #C1FD53E5D4CE971933EC50C9F307AE2171A2D3B52C804642A7A35F84F3A4EA98#)))
#+end_src

#+RESULTS[35fd241b186369c3c59217a2d5f79bd1c1406e61]: cfg-nonguix-pub
#+begin_src text
(mixed-text-file
  "cfg-file"
"(public-key (ecc (curve Ed25519) (q #C1FD53E5D4CE971933EC50C9F307AE2171A2D3B52C804642A7A35F84F3A4EA98#)))")
#+end_src

#+begin_src scheme :noweb-ref system-service
(simple-service
 'nonguix guix-service-type
 (guix-extension
   (substitute-urls (list "https://substitutes.nonguix.org"))
   (authorized-keys (list <<cfg-nonguix-pub()>>))))
#+end_src
** networking
#+begin_src scheme :noweb-ref system-use
(gnu services ssh)
(gnu packages cups)
(gnu services dns)
(gnu services networking)
(gnu services monitoring)
(gnu services avahi)
#+end_src

#+name: cfg-resolv-conf
#+begin_src cfg :wrap src text
name_servers=127.0.1.1
resolv_conf_local_only=no
dnsmasq=true
dnsmasq_conf=/etc/dnsmasq.servers
dnsmasq_resolve=/etc/dnsmasq.resolve
#+end_src

#+RESULTS[59231caa410b09bcc087d62e50b22d12e254386d]: cfg-resolv-conf
#+begin_src text
(mixed-text-file
  "cfg-file"
"name_servers=127.0.1.1\n"
"resolv_conf_local_only=no\n"
"dnsmasq=true\n"
"dnsmasq_conf=/etc/dnsmasq.servers\n"
"dnsmasq_resolve=/etc/dnsmasq.resolve")
#+end_src

#+begin_src scheme :noweb-ref system-service
(service vnstat-service-type)
(service openssh-service-type
         (openssh-configuration
          (x11-forwarding? #t)))
(service dnsmasq-service-type
         (dnsmasq-configuration
          (no-hosts? #t)
          (servers-file "/etc/dnsmasq.servers")
          (listen-addresses '("::1" "127.0.1.1"))
          ;; (addresses '("/dev.local/127.0.0.1"
          ;;              "/local/127.0.0.1"))
          (negative-cache? #f)
          (servers '("9.9.9.9"))
          (extra-options '("--bind-interfaces"
                           "--dns-loop-detect"
                           "--interface=lo"))))
(simple-service
 'resolvconf etc-service-type
 (list `("resolvconf.conf" ,<<cfg-resolv-conf()>>)))

(service ntp-service-type)

(service iwd-service-type
         (iwd-configuration
          (shepherd-provision '(iwd networking wireless-daemon))
          (config
           (iwd-settings
            (general
             (iwd-general-settings
              (enable-network-configuration? #t)))
            (network
             (iwd-network-settings
              (name-resolving-service 'resolvconf)))))))
#+end_src
#+begin_src scheme :noweb-ref guix-pkg
"emacs-iwd-manager"
#+end_src
#+begin_src elisp
(cl-defun iwd-manager--format-mode-line-nerd-icons (&key state ssid scanning &allow-other-keys)
  (concat
   " "
   (cond
    (scanning #("󱛄" 0 1 (display (raise 0.05))))
    ((eq state 'connected) #("󰖩" 0 1 (display (raise 0.05))))
    (t #("󰤮" 0 1 (display (raise 0.05)))))
   (if (and ssid (not (string= ssid "SARG"))) (concat " " ssid " ") " ")))

(use-package iwd-manager
  :defer-incrementally t
  :custom
  (iwd-manager-mode-line-formatter #'iwd-manager--format-mode-line-nerd-icons)
  :config
  (iwd-manager-register-agent)
  (when (modulep! :editor evil +everywhere)
    (evil-collection-inhibit-insert-state 'iwd-manager-mode-map)
    (evil-set-initial-state 'iwd-manager-mode-map 'normal)
    (evil-collection-define-key 'normal 'iwd-manager-mode-map
      "s" 'iwd-manager-scan
      "r" 'revert-buffer
      "p" 'iwd-manager-toggle-power
      "c" 'iwd-manager-connect
      "d" 'iwd-manager-disconnect
      "D" 'iwd-manager-delete-network
      (kbd "RET") 'iwd-manager-connect)))
#+end_src

** vpn
#+begin_src scheme :noweb-ref system-use
(gnu services vpn)
#+end_src
#+begin_src scheme :noweb-ref system-service
(service wireguard-service-type
         (wireguard-configuration
          (auto-start? #f)
          (addresses '("10.66.66.2/32" "fd42:42:42::2/128"))
          (private-key (plain-file "private" (getenv "WG_CLIENT")))
          (peers
           (list
            (wireguard-peer
             (name "hetzner")
             (endpoint "vpn.sarg.org.ru:52817")
             (public-key "6gNRvmvi5oRGSPr8J0dBcyDyKS94zO4Y4Jbwo2u+iV0=")
             (preshared-key (plain-file "psk" (getenv "WG_PSK")))
             (allowed-ips '("0.0.0.0/0" "::/0")))))))
#+end_src
#+begin_src elisp
(defun sarg/vpn-start (country)
  "Start VPN in COUNTRY."
  (interactive (list (completing-read "Country: " '("de" "ru") nil t "de")))
  (message "Starting VPN")
  (sarg/telega-msg-send (telega-chat-by-username "DiogenisBot")
                        (concat "/vpn " country))
  (promise-chain
      (sarg/telega-msg-expect
       (telega-match-gen-predicate 'msg
         '(and (contains "VPN started")
               (chat (bot-user (username "DiogenisBot"))))))
    (then (lambda (msg)
            (eshell-command "sudo herd start wireguard-wg0")

            (when (= 1 (call-process "ip" nil nil nil "a" "show" "dev" "wg0"))
              (notifications-notify :title "Hey" :body "wireguard not running"))
            (notifications-notify :title "Hey" :body (telega-msg-content-text msg))))))


(defun sarg/vpn-stop ()
  "Stop VPN."
  (interactive)
  (unless (= 1 (call-process "pgrep" nil nil nil "aria2c"))
    (user-error "Aria2 is still running"))
  (eshell-command "sudo herd stop wireguard-wg0")
  (sarg/telega-msg-send (telega-chat-by-username "DiogenisBot") "/vpn stop")
  (promise-chain
      (sarg/telega-msg-expect
       (telega-match-gen-predicate 'msg
         '(and (or (contains "Stopped.+VPN")
                   (contains "Nothing to stop"))
               (chat (bot-user (username "DiogenisBot"))))))
    (then (lambda (msg)
            (notifications-notify :title "Hey" :body (telega-msg-content-text msg))))))
#+end_src
** hardware
#+begin_src scheme :noweb-ref system-use
(gnu services base)
(gnu services cups)
(gnu services pm)
#+end_src
#+begin_src scheme :noweb-ref system-pkg
"bluez" "tlp" "libratbag"
#+end_src
#+name: kindle-udev-rule
#+begin_src cfg :wrap "src scheme"
SUBSYSTEM=="net", ACTION=="add", ENV{ID_VENDOR}=="Linux_3.0.35-lab126_with_fsl-usb2-udc", \
  RUN+="🐜(pkg "iproute2")🐜/sbin/ip link set %k up", \
  RUN+="🐜(pkg "iproute2")🐜/sbin/ip add add 192.168.15.243/24 dev %k"
#+end_src

#+RESULTS[0417f331ea1ec0c2a5b1619e91042ae35f3d3e90]: kindle-udev-rule
#+begin_src scheme
(mixed-text-file
  "cfg-file"
"SUBSYSTEM==\"net\", ACTION==\"add\", ENV{ID_VENDOR}==\"Linux_3.0.35-lab126_with_fsl-usb2-udc\", \\\n"
"  RUN+=\""(pkg "iproute2")"/sbin/ip link set %k up\", \\\n"
"  RUN+=\""(pkg "iproute2")"/sbin/ip add add 192.168.15.243/24 dev %k\"")
#+end_src

#+begin_src scheme :noweb-ref system-service
(service fstrim-service-type)
(service power-profiles-daemon-service-type)
(service tlp-service-type
         (tlp-configuration
           (usb-autosuspend? #f)
           (cpu-boost-on-ac? #t)
           (restore-device-state-on-startup? #t)))

(simple-service
 'setkeycodes activation-service-type
 #~(invoke (string-append #$(pkg "kbd") "/bin/setkeycodes")
           "1d" "56"                  ; lctrl->lalt
           "38" "29"                  ; lalt->lctrl
           "3a" "42"))      ; capslock->lshift

(udev-rules-service 'qmk (pkg "qmk-udev-rules"))
(udev-rules-service 'kindle (file->udev-rule "98-kindle-network.rules" <<kindle-udev-rule()>>))
(udev-rules-service 'uaccess (udev-rule "50-uaccess.rules" "ACTION==\"add\", SUBSYSTEM==\"misc\", KERNEL==\"udmabuf\", TAG+=\"uaccess\""))

(simple-service 'ratbagd dbus-root-service-type (list (pkg "libratbag")))

(service sane-service-type
         (sane-configuration
          (backends (list (pkg "sane-airscan")))))

(service cups-service-type
         (cups-configuration
           (web-interface? #t)
           (extensions (list (pkg "cups-filters")))))
#+end_src
** desktop
#+begin_src scheme :noweb-ref system-use
(gnu services guix)
(gnu services desktop)
(gnu services linux)
(gnu services dbus)
#+end_src
#+begin_src scheme :noweb-ref system-service
;; Add polkit rules, so that non-root users in the wheel group can
;; perform administrative tasks (similar to "sudo").
(service polkit-service-type)
polkit-wheel-service
(simple-service 'polkit-wheel-policies polkit-service-type
                (list <<polkit-rule>>))

(service avahi-service-type)
(service dbus-root-service-type)
(service upower-service-type)
(service bluetooth-service-type)
(service udisks-service-type)
(service elogind-service-type
         (elogind-configuration
          (handle-lid-switch-external-power 'suspend)
          (kill-user-processes? #t)))
#+end_src
#+name: cfg-udisks-wheel
#+begin_src cfg :wrap src text
// Allow udisks2 to mount devices without authentication
// for users in the 'wheel' group.
polkit.addRule(function(action, subject) {
    if ((action.id == "org.freedesktop.udisks2.filesystem-mount-system" ||
         action.id == "org.freedesktop.udisks2.filesystem-mount") &&
        subject.isInGroup("wheel")) {
        return polkit.Result.YES;
    }
});
#+end_src

#+RESULTS[6154ce7f960aa34daf1619b7186e33e88c2d3180]: cfg-udisks-wheel
#+begin_src text
(mixed-text-file
  "cfg-file"
"// Allow udisks2 to mount devices without authentication\n"
"// for users in the 'wheel' group.\n"
"polkit.addRule(function(action, subject) {\n"
"    if ((action.id == \"org.freedesktop.udisks2.filesystem-mount-system\" ||\n"
"         action.id == \"org.freedesktop.udisks2.filesystem-mount\") &&\n"
"        subject.isInGroup(\"wheel\")) {\n"
"        return polkit.Result.YES;\n"
"    }\n"
"});")
#+end_src

#+begin_src scheme :noweb-ref polkit-rule
(file-union
   "polkit-udisks-wheel"
   `(("share/polkit-1/rules.d/udisks-wheel.rules"
      ,<<cfg-udisks-wheel()>>)))
#+end_src
** COMMENT virtualization
#+begin_src scheme :noweb-ref system-use
(gnu services virtualization)
#+end_src
#+begin_src scheme :noweb-ref user-group
"kvm"
#+end_src
#+begin_src scheme :noweb-ref polkit-rule
(pkg "spice-gtk")
#+end_src
#+begin_src scheme :noweb-ref system-service
(service libvirt-service-type)
(service virtlog-service-type)
#+end_src
#+begin_src scheme :noweb-ref privileged-program
(privileged-program
 (program (file-append (pkg "spice-gtk") "/libexec/spice-client-glib-usb-acl-helper"))
 (setuid? #t))
#+end_src
** COMMENT docker
#+begin_src scheme :noweb-ref system-use
(gnu services containers)
(gnu system accounts)
#+end_src
#+begin_src scheme :noweb-ref user-group
"cgroup"
#+end_src
#+begin_src scheme :noweb-ref guix-pkg
"podman-compose" "podman-healthcheckd"
#+end_src
#+begin_src scheme :noweb-ref guix-srv
(simple-service
 'podman-healthcheck
 home-shepherd-service-type
 (let ((command #~(list #$(file-append (pkg "podman-healthcheckd") "/bin/podman-healthcheckd")))
       (log-file #~(string-append %user-log-dir "/podman-healthcheckd.log")))
   (list (shepherd-service
           (documentation "Run podman-healthcheckd daemon.")
           (provision '(podman-healthcheckd))
           (modules '((shepherd support)))
           (start #~(make-forkexec-constructor #$command #:log-file #$log-file))
           (stop #~(make-kill-destructor))))))
#+end_src
#+begin_src scheme :noweb-ref system-service
(service iptables-service-type)
(service rootless-podman-service-type
         (rootless-podman-configuration
          (subgids
           (list (subid-range (name "sarg"))))
          (subuids
           (list (subid-range (name "sarg"))))))
#+end_src
** various services
#+begin_src scheme :noweb-ref system-use
(gnu services admin)
(gnu services sysctl)
#+end_src

#+begin_src scheme :noweb-ref system-service
(service pam-limits-service-type
         ;; For Lutris / Wine esync
         (list (pam-limits-entry "*" 'hard 'nofile 524288)))

(simple-service 'sysctl-custom sysctl-service-type
                '(("fs.inotify.max_user_watches" . "524288")))

(service file-database-service-type
         (file-database-configuration
          (schedule #~(calendar-event #:hours '(6) #:minutes '(0)))
          (excluded-directories
           (append '("/home") %default-file-database-excluded-directories))))
#+end_src
* home.scm template
#+begin_src scheme :noweb-ref nil :tangle ../../guix/home.scm :exports none
;; generated from config.org
(use-modules
 (gnu)
 (guix)
 (guix gexp)
 (guix packages)
 (guix git-download)
 (guix build-system copy)
 (guix build utils)
 (guix utils)
 (guix modules)
 (guix transformations)
 (guix channels)
 (gnu home)
 (gnu services)
 (gnu packages)
 (gnu packages admin)
 (gnu packages base)
 (gnu packages gnupg)
 (gnu packages qt)
 (gnu home services)
 (gnu home services syncthing)
 (gnu home services dotfiles)
 (gnu home services shepherd)
 (gnu home services gnupg)
 (gnu home services xdg)
 (gnu home services desktop)
 (gnu home services pm)
 (gnu home services sound)
 (gnu home services shells)
 (gnu home services ssh)
 (gnu home services mail)
 (gnu home services mpv)
 (gnu home services guix)
 (gnu system shadow)
 (gnu services shepherd)
 (gnu services xorg)
 (doomemacs services doomemacs)
 (personal services symlinks)
 (personal services owntracks)
 (personal services screen-locker)
 (personal services utils)
 (personal services supercron)
 (personal services secretsd)
 (personal services backup)
 (personal packages binary)
 (personal packages xlibre)
 (ice-9 match)
 (ice-9 receive)
 (srfi srfi-1)
 (srfi srfi-11)
 (srfi srfi-26))

(define globals <<globals>>)
(define (global x) (assq-ref globals x))

(define (relative-file file)
  (string-append (current-source-directory) "/" file))
(define (pkg x)
  (if (string? x)
      (receive (package out)
          (specification->package+output x)
        (if (string= "out" out) package
            (list package out)))
      x))

(define (flatten . lst)
  "Return a list that recursively concatenates all sub-lists of LST."
  (define (flatten1 head out)
    (if (list? head)
        (fold-right flatten1 out head)
        (cons head out)))
  (fold-right flatten1 '() lst))

(define home-emacs (pkg "emacs-next-pgtk"))

(define (build-with-home-emacs p)
  (package/inherit p
    (arguments
     (substitute-keyword-arguments (package-arguments p)
       ((#:emacs emacs #f) home-emacs)))))


guix/.config/guix/config.scm
(use-modules (gnu)
             (gnu services)
             (gnu packages)
             (guix packages)
             (guix profiles)
             (guix download)
             (guix utils)
             (srfi srfi-1)
             (pkill9 fhs)
             (personal packages xdisorg)
             (nongnu packages linux)
             (nongnu system linux-initrd))

(use-package-modules
 linux ssh android suckless dns
 xdisorg pulseaudio connman xorg)

(use-service-modules
 desktop ssh networking sysctl
 xorg dbus shepherd sound pm)

(define (corrupt-linux freedo version hash)
  (package
    (inherit freedo)
    (name "linux")
    (version version)
    (source (origin
              (method url-fetch)
              (uri (list (string-append "https://www.kernel.org/pub/linux/kernel/v"
                                        (version-major version) ".x/linux-" version ".tar.xz")))
              (sha256 (base32 hash))))
    (home-page "https://www.kernel.org/")
    (synopsis "Linux kernel with nonfree binary blobs included")
    (description
     "The unmodified Linux kernel, including nonfree blobs, for running GuixSD
on hardware which requires nonfree software to function.")))

(define %linux-5.4.21
  (corrupt-linux linux-libre-5.4 "5.4.21"
                 "1yjv8qg47kb4j4jkcpi9z7v07p0vz3gszpmhrfji5866j97748vd"))

(define %my-syslog.conf
  ;; from http://www.linuxfromscratch.org/lfs/view/stable/chapter06/sysklogd.html
  (plain-file "syslog.conf" "
auth,authpriv.* -/var/log/auth.log
*.*;auth,authpriv.none -/var/log/sys.log
daemon.* -/var/log/daemon.log
kern.* -/var/log/kern.log
mail.* -/var/log/mail.log
user.* -/var/log/user.log
#*.emerg *
"))

(define %custom-grub-entry "
menuentry \"Ubuntu 14.04 ISO\" {
    set isofile=\"/home/sarg/Downloads/focal-desktop-amd64.iso\"
    loopback loop ($root)$isofile
    linux (loop)/casper/vmlinuz boot=casper iso-scan/filename=${isofile} quiet splash
    initrd (loop)/casper/initrd
}
")

(define dhclient-enter-hooks "
make_resolv_conf() {
    touch /etc/dnsmasq.servers
    sed -i \"/#dhcp/,+1d;
             $ a #dhcp\\nserver=/${new_domain_name}/${new_domain_name_servers}\\n\" \\
        /etc/dnsmasq.servers

    kill -HUP $(cat /run/dnsmasq.pid)
}
")

(define extrakeys-service-type
  (shepherd-service-type
   'extrakeys
   (lambda (codes)
     (shepherd-service
      (documentation (string-append "Load extra keys (setkeycodes)."))
      (provision '(extrakeys))
      (start #~(lambda _
                 (zero? (system* #$(file-append kbd "/bin/setkeycodes")
                                 #$@codes))))
      (respawn? #f)))))

(define (append-to-computed-file g text)
  #~(begin
      #$g

      (let ((port (open-file #$output "a")))
        (format port #$text)
        (close port))))

(define-public (manifest->packages manifest) ;; copied from guix/scripts/refresh.scm, referring to it with (@@ ...) stopped working for some reason, kept saying it's an unbound variable
  "Return the list of packages in MANIFEST."
  (filter-map (lambda (entry)
                (let ((item (manifest-entry-item entry)))
                  (if (package? item) item #f)))
              (manifest-entries manifest)))


(define fhs-packages ;; list of packages to use for fhs-service
  (manifest->packages
   (specifications->manifest
    (list
     "libmediainfo" "libxscrnsaver" "eudev" "alsa-lib" "libzen"
     "fontconfig" "zlib" "pulseaudio" "libx11"
     "libnet" "libxcb" "libxau" "libxdmcp" "libxxf86vm" "libxext" "libxi"
     "libxrandr" "libxrender" "libxcursor" "libxfixes" "libxinerama" "libxcb"
     "glu" "libdrm" "libxdamage" "expat" "libpciaccess"

     ;; "libxcomposite" "libxtst" "libxaw" "libxt" "libxrandr" "libxext" "libx11"
     ;; "libxfixes" "glib" "gtk+" "gtk+@2" "bzip2" "zlib" "gdk-pixbuf" "libxinerama"
     ;; "libxdamage" "libxcursor" "libxrender" "libxscrnsaver" "libxxf86vm"
     ;; "libxi" "libsm" "libice" "gconf" "freetype" "curl" "nspr" "nss" "fontconfig"
     ;; "cairo" "pango" "expat" "dbus" "cups" "libcap" "sdl2" "libusb" "dbus-glib"
     ;; "atk" "eudev" "network-manager" "pulseaudio" "openal" "alsa-lib" "mesa"
     ;; "libxmu" "libxcb" "glu" "util-linux" "libogg" "libvorbis" "sdl" "sdl2-image"
     ;; "glew" "openssl" "libidn" "tbb" "flac" "freeglut" "libjpeg" "libpng" "libpng@1.2"
     ;; "libsamplerate" "libmikmod" "libtheora" "libtiff" "pixman" "speex" "sdl-image"
     ;; "sdl-ttf" "sdl-mixer" "sdl2-ttf" "sdl2-mixer" "gstreamer" "gst-plugins-base"
     ;; "glu" "libcaca" "libcanberra" "libgcrypt" "libvpx"
     ;;"librsvg" ;; currently requires compiling, but shouldn't, it's being weird
     ;; "libxft" "libvdpau" "gst-plugins-ugly" "libdrm" "xkeyboard-config" "libpciaccess"
     ;;"ffmpeg@3.4" ;; Disable this because test fails at livf-something (373grdi4fc369v2h29g10672whmv0mvb-ffmpeg-3.4.7.drv)
     ;; "libpng" "libgpg-error" "sqlite" "libnotify"

     ;; "fuse" "e2fsprogs" "p11-kit" "xz" "keyutils" "xcb-util-keysyms" "libselinux"
     ;; "ncurses" "jack" "jack2" "vulkan-loader" "at-spi2-atk" "at-spi2-core" "libsigc++"
     ))))



(define* (grub-conf-with-custom-part fn)
  ;; fn returns computed-file
  (lambda* (#:rest r)
    (let ((grubcfg-computed-file (apply fn r)))
      (computed-file
       (computed-file-name grubcfg-computed-file)
       (append-to-computed-file
        (computed-file-gexp grubcfg-computed-file)
        %custom-grub-entry)
       #:options (computed-file-options grubcfg-computed-file)
       ;; #:options '(#:local-build? #t #:substitutable? #f)
       ))))

(define dnsmasq-service-type
  (shepherd-service-type
   'dnsmasq
   (lambda (a)
     (shepherd-service
      (provision '(dnsmasq))
      (requirement '(networking))
      (documentation "Run the dnsmasq DNS server.")
      (start #~(make-forkexec-constructor
                '(#$(file-append dnsmasq "/sbin/dnsmasq")
                  "--keep-in-foreground"
                  "--pid-file=/run/dnsmasq.pid"
                  "--no-hosts"
                  "--local-service"
                  "--address=/dev.local/127.0.0.1"
                  "--address=/local/127.0.0.1"
                  "--servers-file=/etc/dnsmasq.servers"
                  "--no-resolv"
                  "--server=1.1.1.1")
                #:pid-file "/run/dnsmasq.pid"))
      (stop #~(make-kill-destructor))))))

(define wifi-udev-rule
  (udev-rule
   "80-wifi.rules"
   "SUBSYSTEM==\"net\", ACTION==\"add\", DRIVERS==\"iwlwifi\", KERNEL==\"wl*\", NAME=\"wifi\""))

(operating-system
  (kernel %linux-5.4.21)
  (kernel-arguments '("quiet" "loglevel=1" "ipv6.disable=1"))
  (initrd microcode-initrd)
  (initrd-modules (cons "i915" %base-initrd-modules))
  (firmware (cons* iwlwifi-firmware broadcom-bt-firmware %base-firmware))
  (locale "en_GB.utf8")
  (timezone "Europe/Berlin")
  (keyboard-layout (keyboard-layout "us"))
  (bootloader
   (bootloader-configuration
    (bootloader (bootloader
                 (inherit grub-bootloader)
                 (configuration-file-generator
                  (grub-conf-with-custom-part
                   (bootloader-configuration-file-generator grub-bootloader)))))
    (target "/dev/sda")
    (keyboard-layout keyboard-layout)))
  (file-systems
   (cons* (file-system
            (mount-point "/")
            (device
             (uuid "b2c8548b-de1a-4d6c-8ada-2a60dc50e41b"
                   'ext4))
            (type "ext4"))
          %base-file-systems))
  (host-name "thinkpad")

  (groups (cons* (user-group
                  (system? #t)
                  (name "adbusers"))
                 %base-groups))

  (users (cons* (user-account
                 (name "sarg")
                 (comment "Sergey Trofimov")
                 (group "users")
                 (home-directory "/home/sarg")
                 (supplementary-groups
                  '("wheel" "netdev" "audio" "video" "tty" "input" "adbusers")))
                %base-user-accounts))

  (packages
   (cons*
    (specification->package "nss-certs")
    (specification->package "bluez")
    ;; xorg-server
    ;; xf86-input-libinput xf86-video-intel
    brightnessctl
    tlp
    (filter (lambda (p)
              (not (member (package-name p)
                           '("wireless-tools" "info-reader" "nano" "zile"))))
            %base-packages)))

  (services (cons*
             ;; (service connman-service-type)
             (service wpa-supplicant-service-type
                      (wpa-supplicant-configuration
                       (interface "wifi")
                       (config-file "/etc/network/wpa-supplicant.conf")
                       (extra-options '("-Dnl80211"))))

             (service dhcp-client-service-type)
             (extra-special-file "/etc/dhclient-enter-hooks"
                                 (plain-file "dhclient-enter-hooks"
                                             dhclient-enter-hooks))
             (extra-special-file "/etc/dhclient.conf"
                                 (plain-file "dhclient.conf"
                                             "send host-name = gethostname();"))

             (screen-locker-service physlock)
             ;; (screen-locker-service xlockmore "xlock")

             ;; Add polkit rules, so that non-root users in the wheel group can
             ;; perform administrative tasks (similar to "sudo").
             polkit-wheel-service
             (bluetooth-service)

             (udisks-service)
             (service polkit-service-type)
             (elogind-service)

             (service sysctl-service-type
                      (sysctl-configuration
                       (settings '(("kernel.printk" . "2 4 1 7")))))

             (service fhs-binaries-compatibility-service-type
                      (fhs-configuration
                       (lib-packages fhs-packages)))

             (dbus-service)
             x11-socket-directory-service
             (service tlp-service-type
                      (tlp-configuration
                       (restore-device-state-on-startup? #t)))
             (service pulseaudio-service-type
                      (pulseaudio-configuration
                       (daemon-conf '((flat-volumes . no)
                                      (exit-idle-time . -1)))))
             (service alsa-service-type)

             (service openssh-service-type)
             (service dnsmasq-service-type 0)

             ;; remap lctrl->lalt, lalt->lctrl, capslock->lshift
             (service extrakeys-service-type (list "1d" "56" "38" "29" "3a" "42"))

             (modify-services %base-services
               (syslog-service-type
                config =>
                (syslog-configuration
                 (inherit config)
                 (config-file %my-syslog.conf)))

               (udev-service-type
                config =>
                (udev-configuration
                 (inherit config)
                 (rules (cons*
                         brightnessctl
                         android-udev-rules
                         wifi-udev-rule
                         (udev-configuration-rules config)))))))))
